lib/routes/account.js

```diff
@@ -1045,50 +1128,84 @@ module.exports = function (
           .then(
             function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', request, {
-                email: account.email,
-                uid: account.uid,
-                locale: account.locale
-              })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid, 'accountVerify')
+              /**
+               * Logic for account and token verification
+               *
+               * 1) Attempt to use code as tokenVerificationId to verify session.
+               *
+               * 2) An error is thrown if tokenVerificationId does not exist (check to see if email
+               *    verification code) or the tokenVerificationId does not correlate to the
+               *    account uid (damaged linked/spoofed account)
+               *
+               * 3) Verify account email if not already verified.
+               */
+              return db.verifyToken(code, account)
+                .then(function () {
+                  log.event('tokenVerified', request, {
+                    email: account.email,
+                    uid: account.uid,
+                    locale: account.locale,
+                    tokenVerificationId: code
+                  })
+                })
+                .catch(function (err) {
+                  if (butil.buffersAreEqual(request.payload.code, account.emailCode)) {
+                    // The code is just for the account, not for any sessions
+                    return true
+                  } else {
+                    throw error.invalidVerificationCode()
+                  }
 
-              return db.verifyEmail(account)
-                .then(
-                  function() {
-                    // Our post-verification email is very specific to sync,
-                    // so don't send it if we're sure this is not for sync.
-                    // Older clients will not send a 'service' param here
-                    // so we can't always be sure.
-                    if (! service || service === 'sync') {
-                      return mailer.sendPostVerifyEmail(
-                        account.email,
-                        {
-                          acceptLanguage: request.app.acceptLanguage
-                        }
-                      )
-                    }
+                  // Completely unknown code
+                  return err
```

Comment:
Is this dead code?  Both branches in the above `if` statement terminate control flow.